src/holder.rs

```diff
@@ -2,6 +2,52 @@ use crate::sync::atomic::AtomicPtr;
 use crate::{Domain, HazPtrObject, HazPtrRecord};
 use std::sync::atomic::Ordering;
 
+pub struct HazardPointerArray<'domain, F, const N: usize> {
+    domain: &'domain Domain<F>,
+    records: Option<[&'domain HazPtrRecord; N]>,
+}
+
+impl<'domain, F, const N: usize> HazardPointerArray<'domain, F, N> {
+    pub unsafe fn protect<'l, 'o, T>(
+        &'l mut self,
+        mut sources: [&'_ AtomicPtr<T>; N],
```

Comment:
Hmm, I'm not sure this is the API we want for these. My understanding from the folly API is that while you _allocate_ the hazard pointers as an array, you still protect individual `AtomicPtr`s (that is, they don't necessarily _need_ to be in an array). I'd have to look more closely at the API to make sure though. Could you either take a look at what API folly implements here and make sure we match it, or make this non-`pub` for the time being?

Reply:
iirc Folly returns the head of the list, which forces users of the library to traverse the list the way folly does internally. This works in that each pointer can be used individually, but the release_many optimization relies on the values being returned in the same order that they were given (or at least the head and the tail being presented). I found the easiest way to ensure release_many was called with the correct array was to prevent users of the library for having direct access to the array. Perhaps this `protect` method could be renamed `protect_all` and a new `protect` method that takes an index can be created as well

Reply:
I've worked around this by introducing a new type. I'm not super happy about that, as it adds another step between making the Array and protecting a value, but it prevents double-protecting, and it relies on the Array's drop implementation rather than drop for each individual pointer

Reply:
I definitely like `protect_all` as a name for this :+1: 

It'd be really cool if this could take an array of length any `M <= N`, but unfortunately I don't think that's possible to express quite yet.